adagrams/game.py

```diff
@@ -1,11 +1,169 @@
+    the frequency of letters in letter_dict
+    
+    args:
+        letter_dict - a new dictionary of letters
+    """
+    # chr (character), links a number to an ascii character, since
+    # we're only using capital letters, the range is 65-90
+    letter = chr(randint(65,90))
+
+    if letter_dict[letter] > 0:
+        letter_dict[letter] -= 1
+        return letter
+    else:
+        # Recursion, calling this function on itself, basically
+        # saying if I don't find a letter here, run it again
+        return return_letter(letter_dict)
+
 def draw_letters():
-    pass
+    """
+    Returns a list of 10 random letters generated by
+    return_letter
+    """
+    # Copying the dict so it doesn't run off the same one each time
+    # deep copy via dict comprehension, like a for loop but shorter
+    # probably didn't need a deep copy, but want to be safe
+    letter_dict = {letter : value for letter, value in LETTERS.items()}
+
+    # Using helper function & list comprehension
+    new_letters = [return_letter(letter_dict) for _ in range(10)]
+
+    """Below is an alternative way using methods we've learned so far"""
+    # letter_dict = LETTERS.copy()
+    # new_letters = []
+    # letter_choices = list(LETTERS.keys())
+    
+    # while len(new_letters) < 10:
+    #     letter = choice(letter_choices)
+        
+    #     if letter_dict[letter] != 0:
+    #         new_letters.append(letter)
+    #         letter_dict[letter] -= 1
+    #     else:
+    #         continue
+    # return new_letters
+    
+    return new_letters
 
 def uses_available_letters(word, letter_bank):
-    pass
+    """
+    Returns a boolean based on whether the letter_bank
+    has the letters to make the word
+
+    args:
+        word - a word to search 
+        letter_bank - a list of 10 letters
+    """
+    # List comprehension, like a for loop, but shorter
+    letter_list = [letter.lower() for letter in letter_bank]
+    """Below is an alternative way using methods we've learned so far"""
+    # letter_list = []
+
+    # for letter in letter_bank:
+    #     letter_list.append(letter.lower())
+
+    for letter in word.lower():
+        if letter in letter_list:
+            letter_index = letter_list.index(letter)
+            #pop only removes one based on the index
+            letter_list.pop(letter_index)
+        else:
+            return False
+    return True
 
 def score_word(word):
-    pass
+    """
+    Returns a score of a word
+
+    args:
+        word - a word to score
+    """
+    # Ternary statement, like an if statement but shorter
+    total_score = 0 if len(word) < 7 else 8
+    """Below is an alternative way using methods we've learned so far"""
+    # if len(word) < 7:
+    #     total_score = 0
+    # else:
+    #     total_score = 8
+    
+    for letter in word.upper():
+        for score, letters in LETTER_SCORES.items():
+            if letter in letters:
+                total_score += score
```

Comment:
This works fine! We've only just talked about Big O and time/space complexity, but one thing to notice here is we have a 3 nested for loops, even if we haven't explicitly stated `for___ in ___` all three times.

The `if letter in letters` has a loop under the hood. Nothing wrong with 2 nested for loops. At times they are the easier solution. But! we have 3 here, and a little refactoring to our `LETTER_SCORES` dictionary can help.

Let's give each letter a key and a score value. Is there some overhead setting up each key as a letter and its score value, yes.

It would also let us look up a letter's score in "constant" time (O(1)) rather than iterating through a list, which is "linear" time (O(n)) (`letter in ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"]`). Again, we've only just started talking Big O in class, but that's a brief explanation as to why a dictionary is a beneficial data structure we can use to hold data that needs to be frequently looked up.

```python
for letter in word.upper():
    if letter in LETTER_SCORES:
        total_score += LETTER_SCORES[letter]
```
In this way, we've taken O(n^3) (3 nested for loops) to O(n) (one for loop and one constant look up w/ dictionary), which is a lot more efficient